File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/stateful/multi_setup.rs
use specgate::{SpecEvent, spec_operation, spec_setup};

#[derive(Debug, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.multi_setup")]
pub struct Account {
    #[spec_event]
    pub balance: i32,
}

#[spec_setup("transfer", fills = "source", spec = "fixture.multi_setup")]
pub fn make_source() -> Account {
    Account { balance: 100 }
}

#[spec_setup("transfer", fills = "target", spec = "fixture.multi_setup")]
pub fn make_target() -> Account {
    Account { balance: 0 }
}

#[spec_operation("transfer", spec = "fixture.multi_setup")]
pub fn transfer(source: &mut Account, target: &mut Account, amount: i32) {
    source.balance -= amount;
    target.balance += amount;
    observe_balance("source", source);
    observe_balance("target", target);
}

#[spec_operation("observe_balance", spec = "fixture.multi_setup")]
pub fn observe_balance(role: &str, account: &Account) -> i32 {
    assert!(matches!(role, "source" | "target"));
    account.balance
}

#[test]
fn two_setups_fill_distinct_operation_parameters() {
    let mut source = make_source();
    let mut target = make_target();
    transfer(&mut source, &mut target, 30);
    assert_eq!(source.balance, 70);
    assert_eq!(target.balance, 30);
}

File: test/csharp/SpecGate.CtscFixtures/Conformance/Stateful/MultiSetup/MultiSetup.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.Stateful.MultiSetup;

[SpecEvent("Account")]
public sealed class Account
{
    [SpecEvent("balance")]
    public int Balance { get; set; }
}

public static class Transfer
{
    [SpecSetup("transfer", Fills = "source", Spec = "fixture.multi_setup")]
    public static Account MakeSource() => new() { Balance = 100 };

    [SpecSetup("transfer", Fills = "target", Spec = "fixture.multi_setup")]
    public static Account MakeTarget() => new();

    [SpecOperation("transfer", Spec = "fixture.multi_setup")]
    public static void Execute(
        [SpecInput("source")] Account source,
        [SpecInput("target")] Account target,
        [SpecInput("amount")] int amount)
    {
        source.Balance -= amount;
        target.Balance += amount;
        _ = ObserveBalance("source", source);
        _ = ObserveBalance("target", target);
    }

    [SpecOperation("observe_balance", Spec = "fixture.multi_setup")]
    public static int ObserveBalance(
        [SpecInput("role")] string role,
        [SpecInput("account")] Account account)
    {
        ArgumentOutOfRangeException.ThrowIfNotEqual(role is "source" or "target", true);
        return account.Balance;
    }
}
